File: java/anagram/src/main/java/Anagram.java
import java.util.*;
import java.util.stream.*;

class Anagram {
    private String cword;

    Anagram(String word) {
        this.cword = canonize(word);
    }

    List<String> match(List<String> candidates) {
        return candidates
            .stream()
            .filter(this::isAnagram)
            .collect(Collectors.toList());
    }

    private boolean isAnagram(String candidate) {
        return cword.equals(canonize(candidate));
    }

    private static String canonize(String s) {
        return IntStream
            .range(0, s.length())
            .map(
                 i -> i == 0 ?
                 Character.toLowerCase(s.codePointAt(i)) :
                 s.codePointAt(i))
            .sorted()
            .collect(
                     StringBuilder::new,
                     StringBuilder::appendCodePoint,
                     StringBuilder::append)
            .toString();
    }
}

File: kotlin/anagram/src/main/kotlin/Anagram.kt
class Anagram(val source: String) {
    private val charCounts = countChars(source)

    fun match(candidates: List<String>): Set<String> {
        return HashSet(candidates.filter(::isAnagram))
    }

    private fun isAnagram(candidate: String): Boolean {
        return source.toLowerCase() != candidate.toLowerCase() &&
            charCounts == countChars(candidate)
    }
}

private fun countChars(source: String): Map<Char, Int> {
    val result = HashMap<Char, Int>()

    for (c: Char in source.toLowerCase()) {
        val currentCount = result.getOrDefault(c, 0)
        result.put(c, currentCount + 1)
    }

    return result
}
